node/full.go

```diff
@@ -365,59 +360,55 @@ func (n *FullNode) OnStart(ctx context.Context) error {
 	if n.nodeConfig.Aggregator {
 		n.Logger.Info("working in aggregator mode", "block time", n.nodeConfig.BlockTime)
 
-		n.threadManager.Go(func() { n.blockManager.BatchRetrieveLoop(ctx) })
-		n.threadManager.Go(func() { n.blockManager.AggregationLoop(ctx) })
-		n.threadManager.Go(func() { n.blockManager.HeaderSubmissionLoop(ctx) })
-		n.threadManager.Go(func() { n.headerPublishLoop(ctx) })
-		n.threadManager.Go(func() { n.dataPublishLoop(ctx) })
-		return nil
+		go n.blockManager.BatchRetrieveLoop(ctx)
+		go n.blockManager.AggregationLoop(ctx)
+		go n.blockManager.HeaderSubmissionLoop(ctx)
+		go n.headerPublishLoop(ctx)
+		go n.dataPublishLoop(ctx)
+	} else {
+		go n.blockManager.RetrieveLoop(ctx)
+		go n.blockManager.HeaderStoreRetrieveLoop(ctx)
+		go n.blockManager.DataStoreRetrieveLoop(ctx)
+		go n.blockManager.SyncLoop(ctx)
 	}
-	n.threadManager.Go(func() { n.blockManager.RetrieveLoop(ctx) })
-	n.threadManager.Go(func() { n.blockManager.HeaderStoreRetrieveLoop(ctx) })
-	n.threadManager.Go(func() { n.blockManager.DataStoreRetrieveLoop(ctx) })
-	n.threadManager.Go(func() { n.blockManager.SyncLoop(ctx) })
-	return nil
-}
-
-// GetGenesis returns entire genesis doc.
-func (n *FullNode) GetGenesis() *cmtypes.GenesisDoc {
-	return n.genesis
-}
 
-// GetGenesisChunks returns chunked version of genesis.
-func (n *FullNode) GetGenesisChunks() ([]string, error) {
-	err := n.initGenesisChunks()
-	if err != nil {
-		return nil, err
-	}
-	return n.genChunks, err
-}
+	// Block until context is canceled
+	<-ctx.Done()
 
-// OnStop is a part of Service interface.
-//
-// p2pClient and sync services stop first, ceasing network activities. Then rest of services are halted.
-// Context is cancelled to signal goroutines managed by thread manager to stop.
-// Store is closed last because it's used by other services/goroutines.
-func (n *FullNode) OnStop(ctx context.Context) {
+	// Perform cleanup
 	n.Logger.Info("halting full node...")
 	n.Logger.Info("shutting down full node sub services...")
-	err := errors.Join(
+
+	err = errors.Join(
 		n.p2pClient.Close(),
 		n.hSyncService.Stop(ctx),
 		n.dSyncService.Stop(ctx),
 	)
+
 	if n.prometheusSrv != nil {
-		err = errors.Join(err, n.prometheusSrv.Shutdown(ctx))
+		err = errors.Join(err, n.prometheusSrv.Shutdown(context.Background()))
```

Comment:
Shouldn't we pass at least context with timeout here, to be sure that it's not blocked indefinitely?

Reply:
added a context timeout with 2 seconds, nice find